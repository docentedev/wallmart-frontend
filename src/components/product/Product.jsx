import { Plus as PlusIcon, Minus as MinusIcon } from 'react-feather';
import { currencyFormat } from '../../utils'
import styles from './Product.module.css'
import useCart from '../../hooks/useCart';

const Product = ({ product }) => {
    const { get, addItem, substractItem } = useCart()
    const cartItem = get(product.id)

    return (
        <article className={styles.product}>
            <div className={styles.product__image}>
                <img src={product.image} alt={product.description} />
            </div>
            <div className={styles.product__info}>
                <p><strong>{product.brand}</strong> {product.description}</p>
                <strong>{currencyFormat(product.price)}</strong>
            </div>
            <div className={styles.product__actions}>
                {cartItem ? (
                    <div className={styles.product__cart_control}>
                        <button onClick={() => substractItem(product)}>
                            <MinusIcon size={12} />
                        </button>
                        <span>{cartItem.__quantity}</span>
                        <button onClick={() => addItem(product)}>
                            <PlusIcon size={12} />
                        </button>
                    </div>
                ) : (
                        <div>
                            <button
                                className={styles.product__add_button}
                                onClick={() => addItem(product)}
                            >Agregar</button>
                        </div>
                    )}
            </div>
        </article>
    )
}

export default Product
