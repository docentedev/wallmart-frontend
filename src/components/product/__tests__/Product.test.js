import React from 'react'
import ReactDOM from 'react-dom'
import TestRenderer from 'react-test-renderer'
import { act } from 'react-dom/test-utils'

import Product from '../Product'

const objSpy = {
    addActionMock: () => { },
    substractActionMock: () => { },
}

jest.mock('../../../hooks/useCart', () => ({
    __esModule: true,
    default: () => ({
        get: (id) => id === 5 ? ({ __quantity: 2 }) : false,
        addItem: () => {
            objSpy.addActionMock()
        },
        substractItem: () => {
            objSpy.substractActionMock()
        },
    }),
}))

jest.mock('react-feather', () => ({
    __esModule: true,
    Plus: () => (<div className="plus-icon">+</div>),
    Minus: () => (<div className="minus-icon">-</div>),
}))

let container

const product = {
    brand: 'Marca2',
    description: 'Cargador Smart Phone USB',
    id: 5,
    image: 'https://www.lider.cl/catalogo/images/catalogo_no_photo.jpg',
    price: 15000,
    _id: '6000e5dbdd32961e14a9688a',
}

beforeEach(() => {
    container = document.createElement('div')
    document.body.appendChild(container)
})

afterEach(() => {
    document.body.removeChild(container)
    container = null
})

describe('Product test', () => {
    it('should Product render without error', () => {
        const testRenderer = TestRenderer.create(<Product product={product} />)
        const testInstance = testRenderer.root

        expect(testRenderer.toJSON()).toMatchSnapshot()
        expect(testInstance.findByProps({ className: 'plus-icon' }).children).toEqual(['+'])
        expect(testInstance.findByProps({ className: 'minus-icon' }).children).toEqual(['-'])
    })

    it('should call addItem and subsctract functions when the product is in the cart', () => {
        const callAddFunctionSpy = jest.spyOn(objSpy, 'addActionMock')
        const callSubstractcFunctionSpy = jest.spyOn(objSpy, 'substractActionMock')

        act(() => {
            ReactDOM.render(<Product product={product} />, container)
        });
        const buttonAdd = container.querySelector('.add-button')
        const buttonSubstract = container.querySelector('.substract-button')

        act(() => {
            buttonAdd.dispatchEvent(new MouseEvent('click', { bubbles: true }))
        })
        expect(callAddFunctionSpy).toBeCalled()

        act(() => {
            buttonSubstract.dispatchEvent(new MouseEvent('click', { bubbles: true }))
        })
        expect(callSubstractcFunctionSpy).toBeCalled()
    })

    it('should call addItem function when the product is not in the cart', () => {
        product.id = 6
        const callAddFunctionSpy = jest.spyOn(objSpy, 'addActionMock')

        act(() => {
            ReactDOM.render(<Product product={product} />, container)
        });
        const buttonAdd = container.querySelector('button')

        act(() => {
            buttonAdd.dispatchEvent(new MouseEvent('click', { bubbles: true }))
        })
        expect(callAddFunctionSpy).toBeCalled()
    })
})
