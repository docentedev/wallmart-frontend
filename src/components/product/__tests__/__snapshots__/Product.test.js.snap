// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product test should Product render without error 1`] = `
<article
  className="product"
>
  <div
    className="product__image"
  >
    <img
      alt="Cargador Smart Phone USB"
      src="https://www.lider.cl/catalogo/images/catalogo_no_photo.jpg"
    />
  </div>
  <div
    className="product__info"
  >
    <p>
      <strong>
        Marca2
      </strong>
       
      Cargador Smart Phone USB
    </p>
    <strong>
      $15.000
    </strong>
  </div>
  <div
    className="product__actions"
  >
    <div
      className="product__cart_control"
    >
      <button
        className="substract-button"
        onClick={[Function]}
      >
        <div
          className="minus-icon"
        >
          -
        </div>
      </button>
      <span>
        2
      </span>
      <button
        className="add-button"
        onClick={[Function]}
      >
        <div
          className="plus-icon"
        >
          +
        </div>
      </button>
    </div>
  </div>
</article>
`;
